service/controller/resource/tccp/create.go

```diff
@@ -237,52 +240,714 @@ func (r *Resource) getCloudFormationTags(cr v1alpha1.Cluster) []*cloudformation.
+	cc, err := controllercontext.FromContext(ctx)
+	if err != nil {
+		return nil, microerror.Mask(err)
+	}
+
+	var cfg securityConfig
 	{
-		c := adapter.Config{
+		cfg = securityConfig{
 			APIWhitelist:                    r.apiWhiteList,
-			AWSRegion:                       cc.Status.TenantCluster.AWS.Region,
-			ControlPlaneAccountID:           cc.Status.ControlPlane.AWSAccountID,
 			ControlPlaneNATGatewayAddresses: cc.Status.ControlPlane.NATGateway.Addresses,
-			ControlPlanePeerRoleARN:         cc.Status.ControlPlane.PeerRole.ARN,
-			ControlPlaneVPCID:               r.vpcPeerID,
 			ControlPlaneVPCCidr:             cc.Status.ControlPlane.VPC.CIDR,
 			CustomObject:                    cr,
-			EncrypterBackend:                r.encrypterBackend,
-			InstallationName:                r.installationName,
-			PublicRouteTables:               r.publicRouteTables,
-			Route53Enabled:                  r.route53Enabled,
-			StackState: adapter.StackState{
-				Name: key.StackNameTCCP(&cr),
-
-				DockerVolumeResourceName:   key.DockerVolumeResourceName(cr, t),
-				MasterImageID:              key.ImageID(cc.Status.TenantCluster.AWS.Region),
-				MasterInstanceResourceName: key.MasterInstanceResourceName(cr, t),
-				MasterInstanceType:         key.MasterInstanceType(cr),
-				MasterInstanceMonitoring:   r.instanceMonitoring,
-
-				OperatorVersion: key.OperatorVersion(&cr),
+		}
+	}
+
+	masterRules, err := getMasterRules(cfg, cfg.ControlPlaneVPCCidr)
+	if err != nil {
+		return nil, microerror.Mask(err)
+	}
+
+	internalAPIRules, err := getKubernetesPrivateAPIRules(cfg, cfg.ControlPlaneVPCCidr)
+	if err != nil {
+		return nil, microerror.Mask(err)
+	}
+
+	var securityGroups *template.ParamsMainSecurityGroups
+	{
+		securityGroups = &template.ParamsMainSecurityGroups{
+			APIInternalELBSecurityGroupName:  key.SecurityGroupName(&cfg.CustomObject, "internal-api"),
+			APIInternalELBSecurityGroupRules: internalAPIRules,
+			APIWhitelistEnabled:              cfg.APIWhitelist.Public.Enabled,
+			PrivateAPIWhitelistEnabled:       cfg.APIWhitelist.Private.Enabled,
+			MasterSecurityGroupName:          key.SecurityGroupName(&cfg.CustomObject, "master"),
+			MasterSecurityGroupRules:         masterRules,
+			EtcdELBSecurityGroupName:         key.SecurityGroupName(&cfg.CustomObject, "etcd-elb"),
+			EtcdELBSecurityGroupRules:        getEtcdRules(cfg.CustomObject, cfg.ControlPlaneVPCCidr),
+		}
+	}
+
+	return securityGroups, nil
+}
+func newSubnetsParams(ctx context.Context, cr v1alpha1.Cluster) (*template.ParamsMainSubnets, error) {
+	cc, err := controllercontext.FromContext(ctx)
+	if err != nil {
+		return nil, microerror.Mask(err)
+	}
+
+	zones := cc.Spec.TenantCluster.TCCP.AvailabilityZones
+
+	sort.Slice(zones, func(i, j int) bool {
+		return zones[i].Name < zones[j].Name
+	})
+
+	{
+		numAZs := len(zones)
+		if numAZs < 1 {
+			return nil, microerror.Maskf(invalidConfigError, "at least one configured availability zone required")
```

Comment:
This here is backwards. Please check the top of `EnsureCreated`. Also see how other resources like `tcnp` do it. Let's try to validate on the top as a first step. Doing all work and then failing down in the middle eventually is not good. The resource can simply be canceled in the first place when not all information are available yet.